docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,191 @@
+* With TOC:
+  * Get the symbols table and copy it to the local file.
+  * Get the posting table and copy it to the local file.
+  * Get the label offset table and copy it to the local file.
+  * Get the label index and copy it to the local file.
+  
+With that effort building time and resources should be compared with downloading the prebuild `index-header` from bucket. This allows us to reduce complexity of the system as 
+we don't need to cache that in object storage by compactor or even sidecar.
+
+Thanks to this format we can reuse most of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664) code
+to load file. TODO: Benchmark if mmap + loading is actually faster than JSON parsing (3)
+
+This also proposes loading blocks on demand (4).
+
+This means maintaining pool with limited number of `index-header` files loaded at the time in Store Gateway.
+With an LRU logic we should be able to deduce which of blocks should be unloaded and left on the disk. 
+
+This proposes following algorithm:
+
+* blocks newer then X (e.g 2w) being always loaded
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* Background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Unexpected memory usage
+
+Users cares the most about surprising spikes in memory usage. Currently Store Gateway caches the whole index-cache.json. While it's silly to do so for all blocks, this 
+will happen anyway if query will span over large number of blocks and series. This means that while baseline memory will be reduced, baseline vs request memory difference will be even more noticeable.
+
+This tradeoff is acceptable, due to total memory used for all operation should be much smaller. Additionally such query spanning all of the blocks and series are unlikely and should be blocked by simple `sample` limit.
+
+### Benchmark
+
+How to micro-benchmark such change? mmaping is outside of Go run time, which counts allocations etc.
+ 
+### Do we need to mmap?
+
+mmap adds a lot of complexity and confusion especially around monitoring it's memory usage as it does not appear on Go profiles.
+
+While mmap is great for random access against a big file, in fact the current implementation of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664)
+reallocates symbols, offsets and label name=value pairs while reading. This really defies the purpose of mmap as we want to combine all info in a dense few sequential sections of index-header binary format, 
+this file will be mostly read sequentially. Still, label values can be accessed randomly, that's why we propose to start with mmap straight away. 
```

Comment:
But with this, I guess, we would be tied to platforms which support mmap(2). That would exclude Windows AFAIK. It's OK with me, just something to be aware of.

Reply:
In prometheus this seems to work: https://github.com/prometheus/prometheus/blob/master/tsdb/fileutil/mmap_windows.go

Reply:
TIL about that, that's nice! We should definitely re-use that :p disregard this message, then.

Reply:
All is good on this front.